hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/index/IndexResolver.java

```diff
@@ -1191,16 +1129,19 @@ private static IndexFilter composeInFilter(
         List<IndexFilter> newFilters = new ArrayList<>(lastFilter.getFilters().size());
 
         for (IndexFilter filter : lastFilter.getFilters()) {
-            assert filter instanceof IndexEqualsFilter;
+            if (filter instanceof IndexEqualsFilter) {
+                IndexFilter newFilter = composeEqualsFilter(filters, (IndexEqualsFilter) filter, indexType, indexComponentsCount);
 
-            IndexFilter newFilter = composeEqualsFilter(filters, (IndexEqualsFilter) filter, indexType, indexComponentsCount);
+                if (newFilter == null) {
+                    // Cannot create a filter for one of the values of the IN clause. Stop.
+                    return null;
+                }
 
-            if (newFilter == null) {
-                // Cannot create a filter for one of the values of the IN clause. Stop.
-                return null;
+                newFilters.add(newFilter);
+            } else if (filter instanceof IndexRangeFilter) {
+                IndexFilter newFilter = composeRangeFilter(filters, (IndexRangeFilter) filter, indexType, indexComponentsCount);
+                newFilters.add(newFilter);
             }
-
-            newFilters.add(newFilter);
         }
 
         return new IndexInFilter(newFilters);
```

Comment:
Ah, I had that feeling a year ago, and now -- maybe, `IndexInFilter` is not the best name for the class which may contain children filters of different types...

Reply:
Composite it is :+1: